### ISSUE TITLE
Alternating|Tokens and From:To:When

### ISSUE THREAD

**Issue by HaydenReeve** [comfyanonymous/ComfyUI#735]
Hi!

I'd love to enquire about the ability to use two really powerful features from other platforms.

- Prompt Alternating where you have [Human|Duck] and each step it iterates between the token specified.
- Prompt Editing, where it changes based on how many steps completed such as [Photorealistic:Abstract:0.5] where half way through it will change artistic styles.

Thanks!

**Comment by WASasquatch** [comment]
> Hi!
> 
> I'd love to enquire about the ability to use two really powerful features from other platforms.
> 
>     * Prompt Alternating where you have [Human|Duck] and each step it iterates between the token specified.
> 
>     * Prompt Editing, where it changes based on how many steps completed such as [Photorealistic:Abstract:0.5] where half way through it will change artistic styles.
> 
> 
> Thanks!

 - In regards to the first one, that's achieved with `{Human|Duck}` like with sd-dynamic-prompts, and disco diffusion in the past. 
 - I thought to:from:when worked but maybe not? I know `[Photorealistic:0.5]` works.

**Comment by HaydenReeve** [comment]
RE: `{Human|Duck}`

The documentation in the README.md lists this
> You can use `{day|night}`, for wildcard/dynamic prompts. With this syntax `{wild|card|test}` will be randomly replaced by either "wild", "card" or "test" by the frontend every time you queue the prompt. To use {} characters in your actual prompt escape them like: `\{` or `\}`.

This is slightly different functionality from what I am referencing here.

This alternates each time a prompt is queued, not each step of the latent diffusion. The functionality I am describing, at each step, would produce a Human Duck hybrid... Thing. The current functionality of `{Human|Duck}` would generate either a Human or a Duck.

RE: `To:From:When`, this one most certainly doesn't work as expected.

This prompt does not appear to work.
`Man, [Apple:Fire:0.8]` produces
![2023-06-11_19-12-14_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/a5489f3f-c8b8-42be-8418-48e5012292d2)
While this prompt `Man, [Apple::0.8]` produces
![2023-06-11_19-12-37_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/4a530e79-42f4-4833-ac83-25d193ed8f11)

I tried `[Duck|Man:0.2]` and `[Duck|Man:0.8]`
![2023-06-11_19-5-15_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/227221e0-bc75-4a50-b01e-a917d742dbf1)
![2023-06-11_19-5-32_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/1795f4ed-9375-4766-ae46-b84659ea2dc4)

Finally, the `[Photorealistic:0.5]` example also wouldn't work as expected. I tested using an obvious prompt, `Neon`. Here are `Man, `, `Man, [Neon:0.9]`, and `Man, [Neon:0.1]`
![2023-06-11_19-16-55_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/ee276e99-a60e-41e5-8103-0a46c36dd567)
![2023-06-11_19-17-30_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/ba90f0bd-3954-4e67-851d-e92dd5f7f1ba)
![2023-06-11_19-17-46_00001_](https://github.com/comfyanonymous/ComfyUI/assets/39004735/3acfabfc-1fb9-4859-a115-f938a1c5d560)

While they do appear to have an effect on the image, they don't work as a sequencer or as a blend method.

**Comment by WASasquatch** [comment]
Oh to do it for a single diffusion run you have to use stop/start steps with multiple samplers. As for second one, that sucks. Though this works similarly for me `([apple:0.5] [fire:0.5]:1.1)` where both only occur for half the steps.

**Comment by HaydenReeve** [comment]
I assume you're talking about multiple-sampler retaining noise between each? I cannot imagine doing that in an easy fashion. For a 20 step prompt you'd have to have twenty samplers.

I'll have a look at `([apple:0.5] [fire:0.5]:1.1)` but to me that reads as two words; apple and fire, decreased by 50% weighting and then multiplied by 10%. That doesn't exactly seem to be how it works from playing around, but I've not found another explanation that seems to work as expected.

Most surprisingly I've noticed a lot of things that seem to work, but when trying to reduce or test the effects the hypothesis falls apart.

The two features of `{Cat|Dog}` and `{Abstract|Photorealistic:0.8}` are stuff that A1111 does natively, which I was wondering whether they'd be officially supported at some point by Comfy.

**Comment by WASasquatch** [comment]
Brackets isn't decreased weight, it's decreased steps as far as I am aware. Lowering weight is with parenthesis and just using low weight. Brackets control it's occurrence in the diffusion. So 0.5 would be 50% of the steps, so 10 steps. 

The issue with ComfyUI is we encode text early to do stuff with it. Combine, mix, etc, to the input into a sampler already encoded. A1111 has text that it encodes on the fly at diffusion time. So each diffusion step it could parse the text differently.

**Comment by taabata** [comment]
yeah both features are so powerful...if there is a way to implement them in ComfyUI, that would be great.

**Comment by HaydenReeve** [comment]
> Brackets isn't decreased weight, it's decreased steps as far as I am aware. Lowering weight is with parenthesis and just using low weight. Brackets control it's occurrence in the diffusion. So 0.5 would be 50% of the steps, so 10 steps.

That doesn't make sense considering the `Man, [Neon:0.X]` prompt above. the 0.1 should have 10% of the Neon token but it's virtually identical to the `Man, [Neon:0.9]` diffusion.

**Comment by HaydenReeve** [comment]
I am wondering if there would be value in a custom node that does prepares the prompts but does not encode them until a later node, decoupling these functionalities from the node itself. That could allow for the encoding to occur on or before the sampler in custom nodes or officially supported.

I do believe I've found myself wishing more than once to be able to easily append tokens to the prompt without combining the encodings at a later stage for prompt morphing during two/three-resampling stages.

**Comment by WASasquatch** [comment]
> I am wondering if there would be value in a custom node that does prepares the prompts but does not encode them until a later node, decoupling these functionalities from the node itself. That could allow for the encoding to occur on or before the sampler in custom nodes or officially supported.
> 
> I do believe I've found myself wishing more than once to be able to easily append tokens to the prompt without combining the encodings at a later stage for prompt morphing during two/three-resampling stages.

[WAS-NS](https://github.com/WASasquatch/was-node-suite-comfyui) has text editing nodes to setup prompts before putting through a Text to Conditioning or other Conditioning node. There is also Tokens to save your own custom stuff. The NSP condition nodes under WAS Suite/Conditioning allow you to use `<one|two|three>` random prompts which will be reproducible by conditioning seed. It also has prompt variable, so you could do stuff like `$|__color__|$_lights, $1_sign` and it would be parsed to something like `red_lights, red_sign`. 

As far as true to:from:where, I think we need @comfyanonymous to confirm how to do, or if it even exists.

**Comment by WASasquatch** [comment]
Also here is my test

`A man`
![image](https://github.com/comfyanonymous/ComfyUI/assets/1151589/2ebc470f-ec84-4027-b794-b9e361189c28)

`A man, [pink:0.001]`
![image](https://github.com/comfyanonymous/ComfyUI/assets/1151589/08c198ad-7727-4f2e-8e1d-b0b91ddcb656)

`A man, pink`
![image](https://github.com/comfyanonymous/ComfyUI/assets/1151589/f6514b29-f673-4c29-916d-797800c8d922)

`A man, (pink:-1.0)`
![image](https://github.com/comfyanonymous/ComfyUI/assets/1151589/f184efeb-baa4-423c-9ec1-eed59a48c6c5)


To me, it still seems using brackets creates a less defined effect. Pink isn't used for specific things but just in the init_noise, which tricks up the resulting image to incorporate it however, but not defined

**Comment by taabata** [comment]
> Hi!
> 
> I'd love to enquire about the ability to use two really powerful features from other platforms.
> 
> * Prompt Alternating where you have [Human|Duck] and each step it iterates between the token specified.
> * Prompt Editing, where it changes based on how many steps completed such as [Photorealistic:Abstract:0.5] where half way through it will change artistic styles.
> 
> Thanks!

Check out my custom node i created.
https://github.com/taabata/Comfy_custom_nodes
![Screenshot from 2023-06-17 17-13-32](https://github.com/comfyanonymous/ComfyUI/assets/57796911/56d6724e-422c-4a9d-a2be-01d471956abb)

**Comment by SadaleNet** [comment]
I've encountered this exact problem.

And I've just developed a solution for that. The idea is to create a `KSamplerAdvanced` node for each step. Then use a custom `CLIPTextEncodeA1111` node before it that converts A1111-prompt to standard prompt. Then use a textbox to feed the A1111-like prompt to all of the `CLIPTextEncodeA1111`.

Unlike the solution of @taabata, my solution has the potential to support controlnet. However, my solution is messy and requires a lot of nodes (which can be automatically generated using a script included in my repo). The syntax is slightly different form A1111 tho because I don't want to use the `:` as the same character is also used for embedding in ComfyUI. My solution also support recursion syntax.

Here's the repo: [https://github.com/SadaleNet/CLIPTextEncodeA1111-ComfyUI](https://github.com/SadaleNet/CLIPTextEncodeA1111-ComfyUI)

![image](https://github.com/comfyanonymous/ComfyUI/assets/4593277/b917ab1d-614f-4a23-acee-6ba00c906c26)

**Comment by ltdrdata** [comment]
> I've encountered this exact problem.
> 
> And I've just developed a solution for that. The idea is to create a `KSamplerAdvanced` node for each step. Then use a custom `CLIPTextEncodeA1111` node before it that converts A1111-prompt to standard prompt. Then use a textbox to feed the A1111-like prompt to all of the `CLIPTextEncodeA1111`.
> 
> Unlike the solution of @taabata, my solution has the potential to support controlnet. However, my solution is messy and requires a lot of nodes (which can be automatically generated using a script included in my repo). The syntax is slightly different form A1111 tho because I don't want to use the `:` as the same character is also used for embedding in ComfyUI. My solution also support recursion syntax.
> 
> Here's the repo: https://github.com/SadaleNet/CLIPTextEncodeA1111-ComfyUI
> 
> ![image](https://user-images.githubusercontent.com/4593277/257058806-b917ab1d-614f-4a23-acee-6ba00c906c26.png)

Recently someone implemented this. Try this.
https://github.com/asagi4/comfyui-prompt-control

**Comment by coreyryanhanson** [comment]
Rather than having a custom node that tries to do everything at once, or having a ton of different nodes for each step, would it not make sense to have a literal "step" parameter in the k-sampler advanced node? It could function like the third argument in a python range method (start, stop, step) and be called something like "increment" to be less confusing.

You'd be able to achieve the [cat|dog] effect in a more powerful (but more verbose way) using just 2 KSampler (Advanced) nodes that are offset one by in their start step and their respective prompt nodes.

**Comment by WASasquatch** [comment]
> Rather than having a custom node that tries to do everything at once, or having a ton of different nodes for each step, would it not make sense to have a literal "step" parameter in the k-sampler advanced node? It could function like the third argument in a python range method (start, stop, step) and be called something like "increment" to be less confusing.
> 
> You'd be able to achieve the [cat|dog] effect in a more powerful (but more verbose way) using just 2 KSampler (Advanced) nodes that are offset one by in their start step and their respective prompt nodes.

I agree with this logic. Being able to step... the step... would allow you do this elegantly with ksampler advanced. @comfyanonymous does this seem logical?

**Comment by SadaleNet** [comment]
> Rather than having a custom node that tries to do everything at once, or having a ton of different nodes for each step, would it not make sense to have a literal "step" parameter in the k-sampler advanced node? It could function like the third argument in a python range method (start, stop, step) and be called something like "increment" to be less confusing.
> 
> You'd be able to achieve the [cat|dog] effect in a more powerful (but more verbose way) using just 2 KSampler (Advanced) nodes that are offset one by in their start step and their respective prompt nodes.

I don't think this idea would work. It'd require the latent output and latent input of the two KSamplerAdvanced nodes to connect with each others.

**Comment by MoonRide303** [comment]
+1 for that feature. I was often using both alternating words (`[cow|horse]`) and `[from:to:when]` (as well as `[to:when]` and `[from::when]`) syntax to achieve interesting results / transitions in [A1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#prompt-editing) during single sampling pass.
It's an effective way for using different prompts for different steps during sampling, and it would be nice to have it natively supported in ComfyUI. It would probably require enhancing implementation of both CLIP encoders and samplers, though.

**Comment by ltdrdata** [comment]
> +1 for that feature. I was often using both alternating words (`[cow|horse]`) and `[from:to:when]` (as well as `[to:when]` and `[from::when]`) syntax to achieve interesting results / transitions in [A1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#prompt-editing) during single sampling pass. It's an effective way for using different prompts for different steps during sampling, and it would be nice to have it natively supported in ComfyUI. It would probably require enhancing implementation of both CLIP encoders and samplers, though.

Now ComfyUI supports, `ConditioningSetTimestepRange`.

**Comment by WASasquatch** [comment]
> > +1 for that feature. I was often using both alternating words (`[cow|horse]`) and `[from:to:when]` (as well as `[to:when]` and `[from::when]`) syntax to achieve interesting results / transitions in [A1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#prompt-editing) during single sampling pass. It's an effective way for using different prompts for different steps during sampling, and it would be nice to have it natively supported in ComfyUI. It would probably require enhancing implementation of both CLIP encoders and samplers, though.
> 
> Now ComfyUI supports, `ConditioningSetTimestepRange`.

Is there an example of how to do *this* with **that**?

**Comment by MoonRide303** [comment]
> > +1 for that feature. I was often using both alternating words (`[cow|horse]`) and `[from:to:when]` (as well as `[to:when]` and `[from::when]`) syntax to achieve interesting results / transitions in [A1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#prompt-editing) during single sampling pass. It's an effective way for using different prompts for different steps during sampling, and it would be nice to have it natively supported in ComfyUI. It would probably require enhancing implementation of both CLIP encoders and samplers, though.
> 
> Now ComfyUI supports, `ConditioningSetTimestepRange`.

The thing is that for more complex prompts and multiple prompts / CLIP encoders setup we'd be quickly flooded with nodes. Sample (and still relatively simple) prompt from A1111:
```
[dslr photography : oil on canvas painting : 0.1] of a [blue | red] sphere in the city, [dark ink : airbrush : 0.25], dark cyberpunk future, high quality, high resolution
Negative prompt: low quality, low resolution
Steps: 30, Sampler: Euler, CFG scale: 7, Seed: 0, Size: 1024x1024, Model hash: e6bb9ea85b, Model: sd_xl_base_1.0_0.9vae, Clip skip: 2, Score: 7.19, Version: v1.5.1
```
and the output:
![image](https://github.com/comfyanonymous/ComfyUI/assets/130458190/16d41cad-2a1d-44bc-8b57-4b09e2c83ca6)

It's very easy and fun to make that kind of transitions in A1111, and it works pretty well.

Doing something like that via extra nodes would basically mean that for every unique combination of prompt we would have to create duplicates of prompt and conditioning nodes.

And imagine doing it with more advanced flows - for example my basic setup for SDXL is 3 positive + 3 negative prompts (one for each text encoder:  base G+, base G-, base L+, base L-, refiner+, refiner-). If I wanted to do transitions like in the example above in the ComfyUI, I would have to make few time more nodes just to handle that prompt. And each time I would like to add or remove some transitions in the prompt, I would have to reconfigure whole flow.

The prompt2prompt way looks like much better idea to me, to be honest. If anyone would like to (and/or knows how to) implement it in ComfyUI, [here](https://github.com/CompVis/stable-diffusion/commit/3b5c504bb0c11a882252c0eb2b1955474913313a) is original implementation of this feature from Doggettx, and [here](https://github.com/CompVis/stable-diffusion/commit/ccb17b55f2e7acbd1a112b55fb8f8415b4862521) is v2 (might be useful as reference). It would probably work best if it was included in the basic ComfyUI functionality (not as custom nodes).

**Comment by coreyryanhanson** [comment]
> Is there an example of how to do _this_ with **that**? I wasn't getting same sort of results, but I am not exactly sure how to use it, just what seems like how to do it.

I'm pretty sure you can't. It only has start and end. If there was means of stepping through, it seems like it would be the best place for it to avoid the latent piping issue. You'd just set the parameters and add a conditioning (Combine) node. To make more sense the units would probably need to be expressed in steps and then multiple ranges generated on the backend.

Now imagine if that was in place and there was also an advanced recurring "Model (SetTimeStepRange)" It would add additional power beyond what automatic1111 could do.

**Comment by coreyryanhanson** [comment]
> Is there an example of how to do _this_ with **that**? I wasn't getting same sort of results, but I am not exactly sure how to use it, just what seems like how to do it.

For the from:when, you would set the start and end for both prompts and then pipe them into a Conditioning (Combine)

**Comment by taabata** [comment]
The custom node i created allows for token alternating and prompt editing with control net as well. 
link:
https://github.com/taabata/Comfy_Syrian_Falcon_Nodes/tree/main

![Screenshot from 2023-08-06 09-25-42](https://github.com/comfyanonymous/ComfyUI/assets/57796911/fd41e504-ae3d-4012-91e6-786f0c866b54)

**Comment by Lex-DRL** [comment]
I'm late at the party, but +1 for the request.

> Now ComfyUI supports, `ConditioningSetTimestepRange`.

@ltdrdata if I get it right, this node can be used as an alternative to `[from:to:when]` syntax. But:
- It still requires us to manually split text prompt into pieces. What if a prompt contains multiple such entries, each using it's own switch point? This can quickly require literal dozens of nodes just for that.
- As far as I can see, there's still no alternative to `[cow|horse]` syntax. Which is usually used with multiple entries, too. This prompt: `[grey|white|brown] [cow|horse] on a [grass|field|courtyard|lawn|glade]` immediately creates `3*2*5=` 30 prompt variants. Which currently can be achieved only with 30 text node copies and an INSANELY intertwined graph.

Worst of all, both solutions make a network prompt-dependent.

So... is it planned to implement an actual equivalent for this syntax?

**Comment by ltdrdata** [comment]
> I'm late at the party, but +1 for the request.
> 
> > Now ComfyUI supports, `ConditioningSetTimestepRange`.
> 
> @ltdrdata if I get it right, this node can be used as an alternative to `[from:to:when]` syntax. But:
> - It still requires us to manually split text prompt into pieces. What if a prompt contains multiple such entries, each using it's own switch point? This can quickly require literal dozens of nodes just for that.
> - As far as I can see, there's still no alternative to `[cow|horse]` syntax. Which is usually used with multiple entries, too. This prompt: `[grey|white|brown] [cow|horse] on a [grass|field|courtyard|lawn|glade]` immediately creates `3*2*5=` 30 prompt variants. Which currently can be achieved only with 30 text node copies and an INSANELY intertwined graph.
> 
> Worst of all, both solutions make a network prompt-dependent.
> 
> So... is it planned to implement an actual equivalent for this syntax?

Yeah. It seems that we need develope the wrapper.

**Comment by Lex-DRL** [comment]
@ltdrdata 
Maybe, you don't need an edge-case wrapper. Maybe, you need an extension to the current data type.

Sorry if what I'm going to suggest doesn't make sense to you: I'm not sure about the specific python implementation of data flow in ComfyUI. But maybe, a better solution would be something like `conditioning v2` data type, which is treated as an iterator handle?
- unlike current `conditioning`, a dependent nodes gathers the date at each step, not once at evaluations start
- it's the source node's resposibility what it outputs. It may output the same conditioning at each step, but may generate different ones.
- if a current (legacy) datatype is connected, it's just automatically converted into an infinite iterator of the same thing
- To let dependent nodes do any work only once, there could be some metadata attached to indicate the number of unique conditioning objects it generates, their IDs, etc.

**Comment by tbrebant** [comment]
I'm also late to the party too and I will also +1 this request.
I tried the custom nodes presented in this thread.
Sadly @taabata 's one is not working for me. I got different errors that I solved before hitting one that I did not understand.
@SadaleNet 's one works well on my machine but is not scalable.

Prompt alternating is a great way to achieve some effects that are hard to obtain in a different way.

**Comment by comfyanonymous** [comment]
Conditioning concat or combine should give results that are close to alternating prompts.

**Comment by Lex-DRL** [comment]
The issue is, it makes the graph unmanageable. To do it with conditioning concat, we need to manually split a single prompt into multiple nodes... and the split point usually moves within the prompt, which makes the prompting process unnecessary overcomplicated.

**Comment by SadaleNet** [comment]
By adapting the code of the new `ConditioningSetTimestepRange`, it should technically be possible to turn my solution (or other similar solutions) into a scalable one. I don't have the time to do that right now. If someone happen to have implemented a working solution, please do share it here.

**Comment by jesudornenkrone** [comment]
I know that the focus is more on being able to model complex workflows, but being able to write these powerful prompts that easy in A1111 is something I really miss in ComfiyUI.

**Comment by shiimizu** [comment]
> > > +1 for that feature. I was often using both alternating words (`[cow|horse]`) and `[from:to:when]` (as well as `[to:when]` and `[from::when]`) syntax to achieve interesting results / transitions in [A1111](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#prompt-editing) during single sampling pass. It's an effective way for using different prompts for different steps during sampling, and it would be nice to have it natively supported in ComfyUI. It would probably require enhancing implementation of both CLIP encoders and samplers, though.
> > 
> > 
> > Now ComfyUI supports, `ConditioningSetTimestepRange`.
> 
> The thing is that for more complex prompts and multiple prompts / CLIP encoders setup we'd be quickly flooded with nodes. Sample (and still relatively simple) prompt from A1111:
> 
> ```
> [dslr photography : oil on canvas painting : 0.1] of a [blue | red] sphere in the city, [dark ink : airbrush : 0.25], dark cyberpunk future, high quality, high resolution
> Negative prompt: low quality, low resolution
> Steps: 30, Sampler: Euler, CFG scale: 7, Seed: 0, Size: 1024x1024, Model hash: e6bb9ea85b, Model: sd_xl_base_1.0_0.9vae, Clip skip: 2, Score: 7.19, Version: v1.5.1
> ```

This is easily done with my [CLIP Text Encode++](https://github.com/shiimizu/ComfyUI_smZNodes) node.

![ComfyUI_temp_qcebh_00001_](https://github.com/comfyanonymous/ComfyUI/assets/54494639/e1447458-a7a2-466f-9f1d-24d5bb29a5a5)

<img width="1294" alt="workflow-image" src="https://github.com/comfyanonymous/ComfyUI/assets/54494639/bddae45d-bcdb-4952-9af6-1573a6ac648e">

**Comment by MoonRide303** [comment]
@shiimizu Thx for the info, I will check it out.

**Comment by MoonRide303** [comment]
@shiimizu I took quick look at the code, and... I see a lot of sources files copied from A1111. Better idea could be utilising recently introduced 'on prompt' events (#765), and having something more aligned with comfy internals.

**Comment by camoody1** [comment]
@shiimizu I was scanning through this thread tonight looking for a way to use Automatic1111 type syntax in ComfyUI and saw your post about your custom node. I'm just wondering if you're still supporting it, if it's still working, and if you feel like it's a good solution to the discussion here. I feel like these A1111 options are so powerful and it's just crazy to me that Comfy refuses to acknowledge it is lacking in this respect.

Thank you for your time. :)

**Comment by dchatel** [comment]
Another solution might be to get ClipTextEncode node to output a schedule of conditionings rather than a single conditioning. Since we already have the ConditioningSetTimestepRange node, I suppose it wouldn't be that hard to do.

Then the KSampler would get a schedule of conditionings instead of a single one (per positive/negative prompt).
That way, the KSampler could work with something like:
```python
// INPUT: pos_conds, neg_conds where X_conds = [(cond,start,end),...]
for step... :
   current_conditioning = get_the_correct_conditioning_from_X_conds(...) // where X is pos or neg
   do_the_sampling()
```

**Comment by Wyko** [comment]
+1 to this. It's a very important feature for me
